### DISCUSSION TITLE
[Feature request] Use "--color-moved" diffing algorithm in PRs

### DISCUSSION THREAD

**Discussion by MartinSandeCosta** [community/community#9632] [Pull Requests]
PRs that move code around, even within the same function, show `n` lines removed and `n` added. One would assume this diff could be shown as a trivial "code move". This scales poorly when dealing with multiple function trivial code-moves.

This also appears to make contributions on Github a bit misleading: moving code around may give the impression that a certain contributor makes huge number of additions and deletions to a repository, when in fact they didn't contribute to any functionality whatsoever.

As of now, git itself **IS** able to detect such changes: https://git-scm.com/docs/git-diff 
```
--color-moved[=<mode>]
Moved lines of code are colored differently. It can be changed by the diff.colorMoved configuration setting.
```

**Comment by wedamija** [comment]
Please do this. The one thing I miss most about phabricator is that it can differentiate between moved vs actually changed code.

**Reply by brendanwalters** [reply]
I've seen this feature request in a variety of places that aren't github's official feedback (e.g. Twitter) - it's quite popular among those who know about this git feature, which has been around a while and even [appeared in the GitHub blog in 2018](https://github.blog/2018-04-05-git-217-released/).

**Reply by vinkay215** [reply]
I mean the same as you

**Comment by bramjoosten** [comment]
Yup, now I'm discouraged to move code blocks into separate files, because the actual change info is lost with the moving of a code block. even though it's nicer in the long run. No one in my team does this in favor of clear PR's, but we do end up with long files for this reason.

**Reply by vinkay215** [reply]
I mean the same as you

**Comment by 0xdevalias** [comment]
(Crossposted to https://gist.github.com/0xdevalias/1245fe66ac6953d490114b417a0075e3 for posterity)

Omg yes, please! This is one of my biggest pet peeves/code review slowdowns when performing large refactors to cleanup a legacy codebase.

The following were some manual review notes/commands I added to a recent PR to assist others who were reviewing:

> The following `git diff` command will make it easier to see what actually changed in the code here, as opposed to what just moved around without being changed; to make review/etc easier. Dimmed code in the diff has been moved, but not changed.
> 
> This will get all the changes made in this branch, using `devalias/improve-first-use` as the base:
> 
> ```bash
> git \
>   -c color.diff.oldMoved="white dim" \
>   -c color.diff.oldMovedAlternative="white dim" \
>   -c color.diff.newMoved="white dim" \
>   -c color.diff.newMovedAlternative="white dim" \
>   -c color.diff.newMovedDimmed="white dim" \
>   -c color.diff.newMovedAlternativeDimmed="white dim" \
>   diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal \
>   devalias/improve-first-use HEAD
> ```
> 
> This will get all changes made for a specific commit hash (`55a259f`), as compared to the commit hash just before it (`55a259f~`):
> 
> ```bash
> git \
>   -c color.diff.oldMoved="white dim" \
>   -c color.diff.oldMovedAlternative="white dim" \
>   -c color.diff.newMoved="white dim" \
>   -c color.diff.newMovedAlternative="white dim" \
>   -c color.diff.newMovedDimmed="white dim" \
>   -c color.diff.newMovedAlternativeDimmed="white dim" \
>   diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal \
>   55a259f~ 55a259f
> ```
> 
> This will get all changes made for a specific commit hash (`55a259f`), as compared to the commit hash just before it (`55a259f~`), only for the specific file/path specified:
> 
> ```bash
> git \
>   -c color.diff.oldMoved="white dim" \
>   -c color.diff.oldMovedAlternative="white dim" \
>   -c color.diff.newMoved="white dim" \
>   -c color.diff.newMovedAlternative="white dim" \
>   -c color.diff.newMovedDimmed="white dim" \
>   -c color.diff.newMovedAlternativeDimmed="white dim" \
>   diff --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal \
>   55a259f~ 55a259f -- oauth/vendor/slack/index.js
> ```
> 
> You can simplify this for future use by adding it to your `git` aliases in your `~/.gitconfig`, something like the following:
> 
> ```ini
> [alias]
>   diff-refactor = \
>     -c color.diff.oldMoved='white dim' \
>     -c color.diff.oldMovedAlternative='white dim' \
>     -c color.diff.newMoved='white dim' \
>     -c color.diff.newMovedAlternative='white dim' \
>     -c color.diff.newMovedDimmed='white dim' \
>     -c color.diff.newMovedAlternativeDimmed='white dim' \
>     diff --ignore-blank-lines --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal
> ```
> 
> Which would then allow you to use it as just:
> 
> ```bash
> ⇒ git diff-refactor
> 
> ⇒ git diff-refactor devalias/improve-first-use HEAD
> 
> ⇒ git diff-refactor 55a259f~ 55a259f
> 
> ⇒ git diff-refactor 55a259f~ 55a259f -- oauth/vendor/slack/index.js
> ```
> 
> You could potentially add the following as well, though i'm not sure if they will do anything new/useful when combined with `--color-moved`:
> 
> ```bash
> --ignore-all-space --ignore-space-change --ignore-space-at-eol --ignore-cr-at-eol --ignore-blank-lines
> ```

**Reply by 0xdevalias** [reply]
I wonder how hard it would be to take the output from `git diff --color-moved`, parse it, and then render it in a GitHub comment/similar..?

Some potentially useful references:

- https://stackoverflow.com/questions/11509830/how-to-add-color-to-githubs-readme-md-file/73613687#73613687
  - > Now since May 2022, Github can accept [LATEX code on Markdown](https://github.blog/2022-05-19-math-support-in-markdown/), so you can use the `\color{namecolor}` inside the `$$$$`
  - https://github.blog/2022-05-19-math-support-in-markdown/
  - https://en.wikibooks.org/wiki/LaTeX/Colors
- Bash color codes: https://gist.github.com/vratiu/9780109
- https://stackoverflow.com/questions/1380333/highlighting-added-deleted-lines-ignoring-moves-in-a-patch-file
  - https://github.com/l0b0/diff-ignore-moved-lines/blob/master/diff-ignore-moved-lines.sh

---

Super hacky, but this uses `sed` to try and strip out parts of the diff based on the color codes:

```bash
git \
  -c color.diff.oldMoved='white dim' \
  -c color.diff.oldMovedAlternative='white dim' \
  -c color.diff.newMoved='white dim' \
  -c color.diff.newMovedAlternative='white dim' \
  -c color.diff.newMovedDimmed='white dim' \
  -c color.diff.newMovedAlternativeDimmed='white dim' \
  diff --ignore-blank-lines --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal --color HEAD~1 HEAD | sed -r "s/\x1B\[2;37m\+(.*)\x1B\[m/\1/g" | sed -r "s/\x1B\[2m-(.*)\x1B\[m/\1/g" | sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g"
```

Strip 'dim white' `+`:
```bash
sed -r "s/\x1B\[2;37m\+(.*)\x1B\[m/\1/g"
```

Strip 'dim' `-`, keeping the content between:
```bash
sed -r "s/\x1B\[2m-(.*)\x1B\[m/\1/g"
```

Strip all colour codes:
```bash
sed -r "s/\x1B\[([0-9]{1,2}(;[0-9]{1,2})?)?[mGK]//g"
```

If we then put that colour stripped output into a standard GitHub code block with `diff` formatting, it will only use the `+` and `-` lines to determine the colouring to show, and so will highlight the 'actual' changes in red/green.

**Reply by 0xdevalias** [reply]
Another quick little git alias to add to the aliases in my original post ([Ref](https://github.com/orgs/community/discussions/9632#discussioncomment-4702442)):

```ini
diff-refactor-commit = !"f() { git diff-refactor ${1}~1 ${1}; }; f"
```

Original alias for reference:
> You can simplify this for future use by adding it to your `git` aliases in your `~/.gitconfig`, something like the following:
> 
> ```ini
> [alias]
>   diff-refactor = \
>     -c color.diff.oldMoved='white dim' \
>     -c color.diff.oldMovedAlternative='white dim' \
>     -c color.diff.newMoved='white dim' \
>     -c color.diff.newMovedAlternative='white dim' \
>     -c color.diff.newMovedDimmed='white dim' \
>     -c color.diff.newMovedAlternativeDimmed='white dim' \
>     diff --ignore-blank-lines --color-moved=dimmed-zebra --color-moved-ws=ignore-all-space --minimal
> ```
> 
> Which would then allow you to use it as just:
> 
> ```bash
> ⇒ git diff-refactor
> 
> ⇒ git diff-refactor devalias/improve-first-use HEAD
> 
> ⇒ git diff-refactor 55a259f~ 55a259f
> 
> ⇒ git diff-refactor 55a259f~ 55a259f -- oauth/vendor/slack/index.js
> ```

**Reply by 0xdevalias** [reply]
Some similar tangential references to me musing on `--color-moved` type things over the years:

- https://github.com/0xdevalias/chatgpt-source-watch/issues/3#issuecomment-1962228912

> VSCode recently added a new feature called [move code detection](https://code.visualstudio.com/updates/v1_82#_moved-code-detection). The algorithm is surprisingly simple: after normal diff, calculate `similarity(deletion, insertion)` for every `deletion` & `insertion` pair. The code can be found here https://github.com/microsoft/vscode/blob/166097a20cbd06d10d255ef561837c439f372de3/src/vs/editor/common/diff/defaultLinesDiffComputer/computeMovedLines.ts#L44-L85.
> 
> But intuitively demonstrating the moving relationship on TUI might be a challenge.
> 
> _Originally posted by @QuarticCat in https://github.com/Wilfred/difftastic/issues/508#issuecomment-1734691395_

- https://github.com/0xdevalias/chatgpt-source-watch/issues/3#issuecomment-2073984761

Including this one, where I started digging into the git source code to understand how it implements `--color-moved` 'moved detection'; which honestly seems simple enough that it could probably be implemented in a browser extension / user script / similar given the full added/removed diff of a PR:

- https://github.com/0xdevalias/chatgpt-source-watch/issues/3#issuecomment-2542595576

**Reply by AHJiapengZhang** [reply]
> * Explore AST based diff tools 0xdevalias/chatgpt-source-watch#3 (comment)

@0xdevalias Looks like this diff algorithm isn't super hard to implement. The displaying of such a diff might be a bit tricky, though. Because you have to show moved lines across files.

**Reply by Timmmm** [reply]
Phabricator had a simple solution for that. Just highlight the line numbers for moved lines in yellow. When you hover them it has a tooltip saying "Moved from <file>:<line>".

**Reply by AHJiapengZhang** [reply]
> Phabricator had a simple solution for that. Just highlight the line numbers for moved lines in yellow. When you hover them it has a tooltip saying "Moved from :".

@Timmmm This works but it would just nicer to visualize on IDE, especially in the case where a PR consists multiple moving lines between files

**Reply by 0xdevalias** [reply]
Usually during a pull request I don't even care that much where they were moved from; and while it might be nicer to know exactly which file, it's not like `--color-moved` gives that anyway (I don't think); I just end up matching up the 'removed and moved' colour with the 'added and moved' colour; or sometimes I ignore it entirely as I'm usually caring more about the changed implementation details rather than where it moved from/to.

**Reply by 0xdevalias** [reply]
> it would just nicer to visualize on IDE

At that point; just use one of the existing IDE solutions that already supports this? Like VScode?

**Reply by AHJiapengZhang** [reply]
> > it would just nicer to visualize on IDE
> 
> At that point; just use one of the existing IDE solutions that already supports this? Like VScode?

VsCode currently can only visualize moving lines within the same file.

**Reply by AHJiapengZhang** [reply]
> Usually during a pull request I don't even care that much where they were moved from; and while it might be nicer to know exactly which file, it's not like `--color-moved` gives that anyway (I don't think); I just end up matching up the 'removed and moved' colour with the 'added and moved' colour; or sometimes I ignore it entirely as I'm usually caring more about the changed implementation details rather than where it moved from/to.

When the amount of code moved is huge. You want to make sure they are coming from previously used and tested code files and are not tampered accidentally. This is one of the cases I can think where this is useful.

**Comment by soniacasas** [comment]
Any news about this?

**Comment by biajoeknee** [comment]
any news?

**Comment by skwasnia** [comment]
![giphy](https://github.com/community/community/assets/106678034/5ee572e1-7063-44d4-a4d1-9b904cb56791)

**Comment by philippotto** [comment]
As an alternative, you can use VS Code to review while highlighting moved code. It's available behind an experimental flag. It also works in githubs online version. So, simply press `.` in an open PR and review there. Announcement of the feature: https://code.visualstudio.com/updates/v1_82#_moved-code-detection

**Reply by Timmmm** [reply]
That only works within the same file. Phabricator shows lines that are moved between files.

**Reply by AHJiapengZhang** [reply]
Indeed, it doesn't work when functions are moved across files.

**Reply by domainexpansionbankai-boop** [reply]
yup. only works within the same file

**Comment by vinkay215** [comment]
Any news about this?

**Comment by AHJiapengZhang** [comment]
Also vote for this feature. Current Vscode is able to detect line move inside a file. But not so when functions are moved across files.  @hediet

**Comment by biajoeknee** [comment]
you know this isn't being prioritized because the squeaky wheel gets the oil and not enough wheels are squeaking for this.

and the reason not enough wheels are squeaking for this is because the majority of developers are not that scrutinizing (line-by-line) when it comes to code review, so the introduction of this feature wouldn't change many developers user experience much if at all.

**Comment by mberndt123** [comment]
This would be a huge boon for code review.

**Comment by matthewpersico** [comment]
I discovered it this week after using git for 10 years: https://blog.gitbutler.com/how-git-core-devs-configure-git/ . It makes a huge difference when reading a diff.

**Comment by sageserpent-open** [comment]
This doesn’t address the original problem directly, but might be useful if you’re in this situation: https://github.com/sageserpent-open/kineticMerge.

The idea is to merge with awareness of code motion due to refactoring.

Now that is really dealing with the next stage after a PR is reviewed and given the thumbs up, but you could run it with `—no-commit` to see the end result of the merge in a dry-run…

(_I’m the author, BTW_).

**Comment by Judahmeek** [comment]
Come on, Github. Please... 🥺

**Comment by Paul-Hebert** [comment]
This would save a ton of time and make code review and refactors easier! I often comment across lines saying “these lines were moved but unchanged” but this doesn’t scale well at all

**Reply by prdoyle** [reply]
FWIW in that case, I like to use separate commits: have a commit where you only move the lines but don't change them, then change them in another commit.

But yeah... color-moved would be nice.

**Reply by 0xdevalias** [reply]
> have a commit where you only move the lines but don't change them, then change them in another commit.

That works fine when reviewing commit by commit; but when reviewing the full changes in a PR, that doesn't end up helping much.

And while you could potentially argue to make a PR for the moves and another for the changes, in a lot of heavy refactor work that would generally end up being more painful than doing it together.

---

> I often comment across lines saying “these lines were moved but unchanged” but this doesn’t scale well at all

I tend to do similar; and when there are some 'actual changes' and not just 'moved but not really changed', I will often include a screenshot of a local diff that uses `--color-moved` to help show that as well.

**Reply by prdoyle** [reply]
> That works fine when reviewing commit by commit; but when reviewing the full changes in a PR, that doesn't end up helping much.

So... don't do that then?

**Comment by sageserpent-open** [comment]
> This doesn’t address the original problem directly, but might be useful if you’re in this situation: https://github.com/sageserpent-open/kineticMerge.
> 
> The idea is to merge with awareness of code motion due to refactoring.
> 
> Now that is really dealing with the next stage after a PR is reviewed and given the thumbs up, but you could run it with `—no-commit` to see the end result of the merge in a dry-run…
> 
> (_I’m the author, BTW_).

I'm nothing to do with GitHub (well, other than using it to host my own stuff), so I can't help with this feature request at all, but I would be interested to hear from the many participants on this thread as to whether they tried, and had any success with, using Kinetic Merge in `--no-commit` mode to make a dry-run merge.

The idea being to sit on the branch where code has been refactored (which presumably would be the branch being *merged-back*, so to speak, to the mainline in the PR). Kinetic Merge would be used to do this merge from a reversed view point, refreshing the refactoring branch with the latest changes from the mainline.

_(As an aside, I would do this anyway prior to raising a PR for a merge-back, because I want the PR to be purely for review and not for assistance in resolving merge conflicts or getting a build to succeed - that obligation is on me, not the reviewers.)_

Anyway, having the merge done as uncommitted means that `git diff --cached` shows the changes brought *in to* the refactoring branch from further work in the mainline - and it does it without needing to show moved code, because on the refactoring branch, it is already in the right place.

Yes, the moved code is in a different place on the mainline, and that would confuse a standard diff / merge - but Kinetic Merge can cope with that, and what's more, can bring in edits to such moved code made on the mainline, so these edits will show up in the diff in the right place. I don't think Git's moved line detection can cope with edits to moved code, although I'm willing to be corrected on this.

Now this isn't what was orginally asked for, but I'm trying to step back a bit and look at the overall problem here. Maybe this approach could help?

Given I'm the author of said tool, I would certainly be keen to hear of any experiences if people do try it out. 

Probably best to raise tickets over in Kinetic Merge rather than subverting this discussion further with alien concepts. 😁
